corehq/motech/repeaters/tasks.py

```diff
@@ -72,65 +64,112 @@ def delete_old_request_logs():
     check_repeater_lock = get_redis_lock(
-        lock_key,
+        CHECK_REPEATERS_KEY,
         timeout=twentythree_hours_sec,
-        name=lock_key,
+        name=CHECK_REPEATERS_KEY,
     )
     if not check_repeater_lock.acquire(blocking=False):
-        metrics_counter("commcare.repeaters.check.locked_out", tags={'partition': partition})
+        metrics_counter("commcare.repeaters.check.locked_out")
         return
 
     try:
         with metrics_histogram_timer(
             "commcare.repeaters.check.processing",
             timing_buckets=_check_repeaters_buckets,
         ):
-            for record in RepeatRecord.objects.iter_partition(
-                    start, partition, CHECK_REPEATERS_PARTITION_COUNT):
-                if not _soft_assert(
-                    datetime.utcnow() < twentythree_hours_later,
-                    "I've been iterating repeat records for 23 hours. I quit!"
-                ):
-                    break
-
+            for repeater in iter_ready_repeaters():
                 metrics_counter("commcare.repeaters.check.attempt_forward")
-                record.attempt_forward_now(is_retry=True)
-            else:
-                iterating_time = datetime.utcnow() - start
-                _soft_assert(
-                    iterating_time < timedelta(hours=6),
-                    f"It took {iterating_time} to iterate repeat records."
-                )
+                process_repeater.delay(repeater.domain, repeater.repeater_id)
+
+                if datetime.utcnow() > twentythree_hours_later:
+                    # Break after process_repeater() to avoid a stale lock
+                    break
     finally:
         check_repeater_lock.release()
 
 
+def iter_ready_repeaters():
+    """
+    Cycles through repeaters (repeatedly ;) ) until there are no more
+    repeat records ready to be sent.
+    """
+    while True:
+        yielded = False
+        for repeater in Repeater.objects.all_ready():
+            if not repeater.domain_can_forward:
+                continue
+
+            if rate_limit_repeater(repeater.domain):
+                repeater.rate_limit()
+                continue
+
+            yielded = True
+            yield repeater
+
+        if not yielded:
+            # No repeaters are ready, or they are rate limited, or their
+            # domains can't forward or are paused.
+            return
+
+
+@task(queue=settings.CELERY_REPEAT_RECORD_QUEUE)
+def process_repeater(domain, repeater_id):
+
+    def is_retry(repeat_record):
+        return repeat_record.state == State.Fail
+
+    lock_key = f'process_repeater_{repeater_id}'
+    repeater_lock = get_redis_lock(lock_key, timeout=23 * 60 * 60, name=lock_key)
+    if not repeater_lock.acquire(blocking=False):
+        return
+
+    try:
+        repeater = Repeater.objects.get(domain=domain, id=repeater_id)
+        repeat_records = repeater.repeat_records_ready[:repeater.num_workers]
```

Comment:
A bit nit picky since this is much better than the current behavior, but once we call `process_repeat_record`, we have no way of rate limiting at that point right? Having `MAX_REPEATERS` set to 7 or anything on that scale does seem sufficient in preventing a project from hogging repeater resources, but I'm concerned that whoever changes `MAX_REPEATERS` down the line might not take into consideration the impact on rate limiting?

Reply:
Yeah I think so. The alternative would be to include rate limiting code in the `process_repeat_record` method as well right? Just want to understand if you have specific hesitations about taking that step?